File: Tests/server/server_executors/executors/UploadOctetStreamExecutor.cpp
#include "UploadOctetStreamExecutor.h"

#include "Log.h"

void UploadOctetStreamExecutor::doPost(framework::HTTPRequest& request, framework::HTTPResponse& response)
{
	const auto& [data, size, last] = request.getLargeData();

	if (!stream.is_open())
	{
		stream.open(request.getHeaders().at("File-Name"), std::ios::binary);
	}

	if (Log::isValid())
	{
		Log::info("Write {} bytes", "LogUploadOctetStreamExecutor", data.size());
	}

	stream.write(data.data(), data.size());

	if (last)
	{
		if (Log::isValid())
		{
			Log::info("Last upload iteration", "LogUploadOctetStreamExecutor");
		}

		stream.close();

		response.setResponseCode(framework::ResponseCodes::created);
		response.setBody("Finish uploading file");
	}
}

DECLARE_EXECUTOR(UploadOctetStreamExecutor)

File: Tests/executors/CSharp/CoreExecutors/UploadOctetStreamExecutor.cs
using Framework;
using Framework.Utility;

public class UploadOctetStreamExecutor : StatefulExecutor
{
	private FileStream? stream;

	public override void DoPost(HttpRequest request, HttpResponse response)
	{
		var (data, last) = request.GetLargeData();

		stream ??= new(request.GetHeaders()["File-Name"], FileMode.Create);

		stream.Write(data);

		if (last)
		{
			stream.Dispose();

			stream = null;

			response.SetResponseCode(ResponseCodes.Created);
			response.SetBody("Finish uploading file");
		}
	}
}
